streams-tests/jvm/src/test/scala/zio/stream/SinkSpec.scala

```diff
@@ -17,29 +17,791 @@ class SinkSpec(implicit ee: org.specs2.concurrent.ExecutionEnv)
+
+    foldWeighted $foldWeighted
+
+    foldWeightedM $foldWeightedM
+
+    foldUntil $foldUntil
+
+    foldUntilM $foldUntilM
+
+    fromFunction $fromFunction
+
+    fromOutputStream $fromOutputStream
+
+    throttleEnforce $throttleEnforce
+      with burst    $throttleEnforceWithBurst
+
+    throttleShape        $throttleShape
+      infinite bandwidth $throttleShapeInfiniteBandwidth
+      with burst         $throttleShapeWithBurst
 
   Usecases
     Number array parsing with Sink.foldM  $jsonNumArrayParsingSinkFoldM
     Number array parsing with combinators $jsonNumArrayParsingSinkWithCombinators
   """
 
+  private def initErrorSink = new ZSink[Any, String, Int, Int, Int] {
+    type State = Unit
+    val initial                    = IO.fail("Ouch")
+    def step(state: State, a: Int) = IO.fail("Ouch")
+    def extract(state: State)      = IO.fail("Ouch")
+  }
+
+  private def stepErrorSink = new ZSink[Any, String, Int, Int, Int] {
+    type State = Unit
+    val initial                    = UIO.succeed(Step.more(()))
+    def step(state: State, a: Int) = IO.fail("Ouch")
+    def extract(state: State)      = IO.fail("Ouch")
+  }
+
+  private def extractErrorSink = new ZSink[Any, String, Int, Int, Int] {
+    type State = Unit
+    val initial                    = UIO.succeed(Step.more(()))
+    def step(state: State, a: Int) = UIO.succeed(Step.done((), Chunk.empty))
+    def extract(state: State)      = IO.fail("Ouch")
+  }
+
+  private def sinkIteration[R, E, A0, A, B](sink: ZSink[R, E, A0, A, B], a: A) =
+    for {
+      init   <- sink.initial
+      step   <- sink.step(Step.state(init), a)
+      result <- sink.extract(Step.state(step))
+    } yield result
+
+  private def chunkedHappyPath = {
+    val sink = ZSink.collectAll[Int].chunked
+    unsafeRun(sinkIteration(sink, Chunk(1, 2, 3, 4, 5)).map(_ must_=== List(1, 2, 3, 4, 5)))
+  }
+
+  private def chunkedEmpty = {
+    val sink = ZSink.collectAll[Int].chunked
+    unsafeRun(sinkIteration(sink, Chunk.empty).map(_ must_=== Nil))
+  }
+
+  private def chunkedInitError = {
+    val sink = initErrorSink.chunked
+    unsafeRun(sinkIteration(sink, Chunk.single(1)).either.map(_ must_=== Left("Ouch")))
+  }
+
+  private def chunkedStepError = {
+    val sink = stepErrorSink.chunked
+    unsafeRun(sinkIteration(sink, Chunk.single(1)).either.map(_ must_=== Left("Ouch")))
+  }
+
+  private def chunkedExtractError = {
+    val sink = extractErrorSink.chunked
+    unsafeRun(sinkIteration(sink, Chunk.single(1)).either.map(_ must_=== Left("Ouch")))
+  }
+
+  private def collectAllHappyPath = {
+    val sink = ZSink.identity[Int].collectAll[Int, Int]
+    unsafeRun(sinkIteration(sink, 1).map(_ must_=== List(1)))
+  }
+
+  private def collectAllInitError = {
+    val sink = initErrorSink.collectAll
+    unsafeRun(sinkIteration(sink, 1).either.map(_ must_=== Left("Ouch")))
+  }
+
+  private def collectAllStepError = {
+    // This test needs to be verified for correctness.
```

Comment:
Yeah I think the implementation for `Sink#collectAllWith` is surprising. It stores the error in the first element of the tuple, but never uses it in `extract`; and if subsequent errors occur in a step, the previous errors are discarded. We should probably change it to fail eagerly in `step`.